File: test/services/task_outcome_test.dart
import 'dart:math';
import 'package:flutter_test/flutter_test.dart';
import 'package:pixelcode/services/task_outcome.dart';

void main() {
  group('bugChance', () {
    test('clamps to [0, 0.4]', () {
      expect(bugChance(precisionSkill: 0), 0.4);
      expect(bugChance(precisionSkill: 14), 0.0);
      expect(bugChance(precisionSkill: 7), closeTo(0.19, 1e-9));
      expect(bugChance(precisionSkill: 100), 0.0);
    });
  });

  group('critChance', () {
    test('is 0.02 * creativity', () {
      expect(critChance(creativitySkill: 0), 0.0);
      expect(critChance(creativitySkill: 10), closeTo(0.2, 1e-9));
    });

    test('clamps at 1.0', () {
      expect(critChance(creativitySkill: 100), 1.0);
    });
  });

  group('completionSuccessChance', () {
    test('starts at 0.85', () {
      expect(completionSuccessChance(reliabilitySkill: 0), 0.85);
    });

    test('caps at 1.0', () {
      expect(completionSuccessChance(reliabilitySkill: 100), 1.0);
    });

    test('reaches 1.0 by reliability 15', () {
      expect(completionSuccessChance(reliabilitySkill: 15), 1.0);
    });
  });

  group('rollOutcome', () {
    test('high-everything agent almost always gets clean', () {
      final rng = Random(1);
      var clean = 0;
      var incomplete = 0;
      for (var i = 0; i < 1000; i++) {
        final r = rollOutcome(
          rng: rng,
          precisionSkill: 10,
          creativitySkill: 0,
          reliabilitySkill: 15,
        );
        if (r == TaskOutcome.clean) clean++;
        if (r == TaskOutcome.incomplete) incomplete++;
      }
      expect(clean, greaterThan(850));
      expect(incomplete, 0);
    });

    test('low-reliability agent often incomplete', () {
      final rng = Random(1);
      var incomplete = 0;
      for (var i = 0; i < 1000; i++) {
        final r = rollOutcome(
          rng: rng,
          precisionSkill: 10,
          creativitySkill: 0,
          reliabilitySkill: 0,
        );
        if (r == TaskOutcome.incomplete) incomplete++;
      }
      // 0.85 success rate → ~15% incomplete
      expect(incomplete, greaterThan(100));
      expect(incomplete, lessThan(200));
    });

    test('low-precision agent often bugs', () {
      final rng = Random(1);
      var bugs = 0;
      for (var i = 0; i < 1000; i++) {
        final r = rollOutcome(
          rng: rng,
          precisionSkill: 0,
          creativitySkill: 0,
          reliabilitySkill: 15,
        );
        if (r == TaskOutcome.bug) bugs++;
      }
      // bug chance 0.4 × success 1.0 → ~40%
      expect(bugs, greaterThan(350));
      expect(bugs, lessThan(450));
    });

    test('crit only fires on divergent tasks', () {
      final rng = Random(1);
      var critsNonDiv = 0;
      for (var i = 0; i < 500; i++) {
        final r = rollOutcome(
          rng: rng,
          precisionSkill: 15,
          creativitySkill: 50,
          reliabilitySkill: 15,
          isDivergentTask: false,
        );
        if (r == TaskOutcome.crit) critsNonDiv++;
      }
      expect(critsNonDiv, 0);

      var critsDiv = 0;
      for (var i = 0; i < 500; i++) {
        final r = rollOutcome(
          rng: rng,
          precisionSkill: 15,
          creativitySkill: 50,
          reliabilitySkill: 15,
          isDivergentTask: true,
        );
        if (r == TaskOutcome.crit) critsDiv++;
      }
      expect(critsDiv, greaterThan(0));
    });

    test('divergentTaskTypes covers expected IDs', () {
      expect(divergentTaskTypes, containsAll({'architecture', 'product-spec', 'ui-design'}));
    });

    test('workstationTaskTypes covers coding/testing/debugging', () {
      expect(workstationTaskTypes, containsAll({'coding', 'testing', 'debugging'}));
    });

    test('unassigned agent on coding task has higher incomplete rate', () {
      final rng = Random(42);
      var incompleteAssigned = 0;
      var incompleteUnassigned = 0;
      const n = 2000;
      for (var i = 0; i < n; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
              taskType: 'coding',
              isUnassigned: false,
            ) ==
            TaskOutcome.incomplete) {
          incompleteAssigned++;
        }
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
              taskType: 'coding',
              isUnassigned: true,
            ) ==
            TaskOutcome.incomplete) {
          incompleteUnassigned++;
        }
      }
      // Unassigned penalty (−0.25 success) must produce more incompletes.
      expect(incompleteUnassigned, greaterThan(incompleteAssigned));
    });

    test('unassigned penalty does NOT apply on non-workstation tasks', () {
      // For a task like "architecture", isUnassigned should have no extra effect.
      final rng = Random(7);
      var incompleteAssigned = 0;
      var incompleteUnassigned = 0;
      const n = 2000;
      for (var i = 0; i < n; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
              taskType: 'architecture',
              isUnassigned: false,
            ) ==
            TaskOutcome.incomplete) {
          incompleteAssigned++;
        }
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
              taskType: 'architecture',
              isUnassigned: true,
            ) ==
            TaskOutcome.incomplete) {
          incompleteUnassigned++;
        }
      }
      // Without the penalty the two distributions should be within normal
      // statistical noise (~5% tolerance on 2000 samples).
      final diff = (incompleteUnassigned - incompleteAssigned).abs();
      expect(diff, lessThan(n * 0.05));
    });

    test('high-reliability unassigned agent on coding task still completes', () {
      // reliabilitySkill 15 → success 1.0; penalty −0.25 → 0.75 still > 0.
      final rng = Random(99);
      var incomplete = 0;
      for (var i = 0; i < 1000; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 15,
              taskType: 'coding',
              isUnassigned: true,
            ) ==
            TaskOutcome.incomplete) {
          incomplete++;
        }
      }
      // ~25% incomplete expected, not 0 and not all.
      expect(incomplete, greaterThan(180));
      expect(incomplete, lessThan(320));
    });

    test('specialization bonus raises crit rate on divergent tasks', () {
      final rng = Random(11);
      var crits = 0;
      var critsBaseline = 0;
      const n = 4000;
      for (var i = 0; i < n; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 15,
              creativitySkill: 5,
              reliabilitySkill: 15,
              isDivergentTask: true,
              specializationCritBonus: 0.15,
            ) ==
            TaskOutcome.crit) {
          crits++;
        }
        if (rollOutcome(
              rng: rng,
              precisionSkill: 15,
              creativitySkill: 5,
              reliabilitySkill: 15,
              isDivergentTask: true,
            ) ==
            TaskOutcome.crit) {
          critsBaseline++;
        }
      }
      // base critChance(5) = 0.10; with +0.15 bonus → 0.25.
      // Bonus run must be meaningfully higher (delta well above noise).
      expect(crits - critsBaseline, greaterThan(n * 0.08));
    });

    test('specialization bonus does NOT fire on non-divergent tasks', () {
      final rng = Random(13);
      var crits = 0;
      for (var i = 0; i < 1000; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 15,
              creativitySkill: 50,
              reliabilitySkill: 15,
              isDivergentTask: false,
              specializationCritBonus: 0.30,
            ) ==
            TaskOutcome.crit) {
          crits++;
        }
      }
      expect(crits, 0);
    });

    test('specialization bonus is capped at kMaxSpecializationCritBonusInRoll',
        () {
      // Pass an absurd bonus; effective contribution must equal the cap.
      // base critChance(0) = 0; with cap = 0.30 → ~30% crit on divergent.
      final rng = Random(17);
      var crits = 0;
      const n = 4000;
      for (var i = 0; i < n; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 15,
              creativitySkill: 0,
              reliabilitySkill: 15,
              isDivergentTask: true,
              specializationCritBonus: 5.0, // wildly above the cap
            ) ==
            TaskOutcome.crit) {
          crits++;
        }
      }
      // Should sit around 30% (the cap), never near 100%.
      expect(crits, greaterThan(n * 0.22));
      expect(crits, lessThan(n * 0.38));
    });
  });

  group('lessonSuccessBonus', () {
    test('zero lessons yields no bonus', () {
      expect(lessonSuccessBonus(lessonCount: 0), 0.0);
    });

    test('scales at 0.005 per lesson', () {
      expect(lessonSuccessBonus(lessonCount: 10), closeTo(0.05, 1e-9));
      expect(lessonSuccessBonus(lessonCount: 20), closeTo(0.10, 1e-9));
    });

    test('caps at kMaxLessonSuccessBonusInRoll', () {
      expect(lessonSuccessBonus(lessonCount: 100), kMaxLessonSuccessBonusInRoll);
    });
  });

  group('rollOutcome — lesson bonus', () {
    test('lesson bonus reduces incomplete rate', () {
      final rng = Random(23);
      var incompleteNoLessons = 0;
      var incompleteMaxLessons = 0;
      const n = 4000;
      for (var i = 0; i < n; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
            ) ==
            TaskOutcome.incomplete) {
          incompleteNoLessons++;
        }
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
              lessonBonus: kMaxLessonSuccessBonusInRoll,
            ) ==
            TaskOutcome.incomplete) {
          incompleteMaxLessons++;
        }
      }
      // With max lesson bonus the agent should have meaningfully fewer incompletes.
      expect(incompleteNoLessons, greaterThan(incompleteMaxLessons));
      expect(incompleteNoLessons - incompleteMaxLessons, greaterThan(n * 0.04));
    });

    test('lesson bonus is capped at kMaxLessonSuccessBonusInRoll', () {
      // reliability=0 → base 0.85; cap adds at most +0.10 → 0.95 success.
      // Incomplete rate ~5%, never near 0% (uncapped) or 15% (no bonus).
      final rng = Random(29);
      var incomplete = 0;
      const n = 4000;
      for (var i = 0; i < n; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 10,
              creativitySkill: 0,
              reliabilitySkill: 0,
              lessonBonus: 99.0,
            ) ==
            TaskOutcome.incomplete) {
          incomplete++;
        }
      }
      expect(incomplete, greaterThan(n * 0.02));
      expect(incomplete, lessThan(n * 0.10));
    });

    test('lesson bonus does NOT produce crits on non-divergent tasks', () {
      final rng = Random(37);
      var crits = 0;
      for (var i = 0; i < 1000; i++) {
        if (rollOutcome(
              rng: rng,
              precisionSkill: 0,
              creativitySkill: 50,
              reliabilitySkill: 15,
              isDivergentTask: false,
              lessonBonus: kMaxLessonSuccessBonusInRoll,
            ) ==
            TaskOutcome.crit) {
          crits++;
        }
      }
      expect(crits, 0);
    });
  });
}

File: server/test/task_outcome.test.ts
import { test, describe } from "node:test";
import assert from "node:assert/strict";

import {
  bugChance,
  completionSuccessChance,
  critChance,
  divergentTaskTypes,
  lessonSuccessBonus,
  mulberry32,
  projectMemoryDepthBonus,
  rollOutcome,
  workstationTaskTypes,
  kMaxLessonSuccessBonusInRoll,
  kMaxProjectMemoryBonusInRoll,
  kMaxSpecializationCritBonusInRoll,
  kUnassignedIncompletePenalty,
  type Rng,
} from "../src/task_outcome.js";

function approx(actual: number, expected: number, eps = 1e-9): void {
  assert.ok(
    Math.abs(actual - expected) < eps,
    `expected ${expected} ± ${eps}, got ${actual}`,
  );
}

// ─── Skill curves (mirror Dart suite) ─────────────────────────────────────────

describe("bugChance", () => {
  test("clamps to [0, 0.4]", () => {
    assert.equal(bugChance(0), 0.4);
    assert.equal(bugChance(14), 0.0);
    approx(bugChance(7), 0.19);
    assert.equal(bugChance(100), 0.0);
  });
});

describe("critChance", () => {
  test("is 0.02 * creativity", () => {
    assert.equal(critChance(0), 0.0);
    approx(critChance(10), 0.2);
  });
  test("clamps at 1.0", () => {
    assert.equal(critChance(100), 1.0);
  });
});

describe("completionSuccessChance", () => {
  test("starts at 0.85", () => {
    assert.equal(completionSuccessChance(0), 0.85);
  });
  test("caps at 1.0", () => {
    assert.equal(completionSuccessChance(100), 1.0);
  });
  test("reaches 1.0 by reliability 15", () => {
    assert.equal(completionSuccessChance(15), 1.0);
  });
});

describe("lessonSuccessBonus", () => {
  test("clamps at kMaxLessonSuccessBonusInRoll (0.10)", () => {
    assert.equal(lessonSuccessBonus(0), 0.0);
    approx(lessonSuccessBonus(10), 0.05);
    assert.equal(lessonSuccessBonus(20), kMaxLessonSuccessBonusInRoll);
    assert.equal(lessonSuccessBonus(1000), kMaxLessonSuccessBonusInRoll);
  });
});

describe("projectMemoryDepthBonus", () => {
  test("clamps at kMaxProjectMemoryBonusInRoll (0.15)", () => {
    assert.equal(projectMemoryDepthBonus(0), 0.0);
    approx(projectMemoryDepthBonus(5), 0.01);
    assert.equal(projectMemoryDepthBonus(75), kMaxProjectMemoryBonusInRoll);
    assert.equal(projectMemoryDepthBonus(10000), kMaxProjectMemoryBonusInRoll);
  });
});

describe("task type sets", () => {
  test("divergentTaskTypes covers architecture/product-spec/ui-design", () => {
    assert.ok(divergentTaskTypes.has("architecture"));
    assert.ok(divergentTaskTypes.has("product-spec"));
    assert.ok(divergentTaskTypes.has("ui-design"));
    assert.equal(divergentTaskTypes.has("coding"), false);
  });
  test("workstationTaskTypes covers coding/testing/debugging", () => {
    assert.ok(workstationTaskTypes.has("coding"));
    assert.ok(workstationTaskTypes.has("testing"));
    assert.ok(workstationTaskTypes.has("debugging"));
    assert.equal(workstationTaskTypes.has("architecture"), false);
  });
});

// ─── rollOutcome — distributional / behavioral ────────────────────────────────

describe("rollOutcome — happy-path agent", () => {
  test("high-everything agent almost always gets clean", () => {
    const rng = mulberry32(1);
    let clean = 0;
    let incomplete = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 10,
        creativitySkill: 0,
        reliabilitySkill: 15,
      });
      if (r === "clean") clean++;
      if (r === "incomplete") incomplete++;
    }
    assert.ok(clean > 850, `clean=${clean}`);
    assert.equal(incomplete, 0);
  });
});

describe("rollOutcome — low-reliability agent", () => {
  test("often incomplete", () => {
    const rng = mulberry32(1);
    let incomplete = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 10,
        creativitySkill: 0,
        reliabilitySkill: 0,
      });
      if (r === "incomplete") incomplete++;
    }
    // 0.85 success rate → ~15% incomplete
    assert.ok(incomplete > 100, `incomplete=${incomplete}`);
    assert.ok(incomplete < 200, `incomplete=${incomplete}`);
  });
});

describe("rollOutcome — low-precision agent", () => {
  test("often bugs", () => {
    const rng = mulberry32(1);
    let bugs = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 0,
        creativitySkill: 0,
        reliabilitySkill: 15,
      });
      if (r === "bug") bugs++;
    }
    // 0.4 bug rate
    assert.ok(bugs > 350, `bugs=${bugs}`);
    assert.ok(bugs < 450, `bugs=${bugs}`);
  });
});

describe("rollOutcome — crit gating", () => {
  test("never crits on convergent task type", () => {
    const rng = mulberry32(42);
    let crit = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 15,
        creativitySkill: 50,
        reliabilitySkill: 15,
        isDivergentTask: false,
      });
      if (r === "crit") crit++;
    }
    assert.equal(crit, 0);
  });

  test("crits on divergent task with high creativity", () => {
    const rng = mulberry32(42);
    let crit = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 15,
        creativitySkill: 50,
        reliabilitySkill: 15,
        isDivergentTask: true,
      });
      if (r === "crit") crit++;
    }
    // creativity 50 → cap 1.0 → essentially always crits
    assert.ok(crit > 900, `crit=${crit}`);
  });

  test("specializationCritBonus is clamped to 0.30", () => {
    const rng = mulberry32(7);
    let crit = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 15,
        isDivergentTask: true,
        specializationCritBonus: 1.0, // clamps to 0.30
      });
      if (r === "crit") crit++;
    }
    // crit chance = 0 + 0.30 = 0.30 → ~300 of 1000
    assert.ok(crit > 230, `crit=${crit}`);
    assert.ok(crit < 370, `crit=${crit}`);
  });

  test("projectMemoryBonus only applies to architecture", () => {
    const rngArch = mulberry32(9);
    const rngSpec = mulberry32(9);
    let critArch = 0;
    let critSpec = 0;
    for (let i = 0; i < 1000; i++) {
      const rArch = rollOutcome({
        rng: rngArch,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 15,
        isDivergentTask: true,
        taskType: "architecture",
        projectMemoryBonus: 1.0, // clamped to 0.15
      });
      const rSpec = rollOutcome({
        rng: rngSpec,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 15,
        isDivergentTask: true,
        taskType: "product-spec",
        projectMemoryBonus: 1.0, // ignored — not architecture
      });
      if (rArch === "crit") critArch++;
      if (rSpec === "crit") critSpec++;
    }
    assert.ok(critArch > 100, `critArch=${critArch}`);
    assert.equal(critSpec, 0, `critSpec=${critSpec}`);
  });
});

describe("rollOutcome — workstation desk penalty", () => {
  test("unassigned coder hits −0.25 success penalty", () => {
    const rngWith = mulberry32(11);
    const rngWithout = mulberry32(11);
    let incompleteWith = 0;
    let incompleteWithout = 0;
    for (let i = 0; i < 1000; i++) {
      const rWith = rollOutcome({
        rng: rngWith,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 0,
        taskType: "coding",
        isUnassigned: false,
      });
      const rWithout = rollOutcome({
        rng: rngWithout,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 0,
        taskType: "coding",
        isUnassigned: true,
      });
      if (rWith === "incomplete") incompleteWith++;
      if (rWithout === "incomplete") incompleteWithout++;
    }
    // base success 0.85 → 15% incomplete; with penalty 0.85−0.25=0.60 → 40% incomplete
    approx(incompleteWith / 1000, 0.15, 0.05);
    approx(incompleteWithout / 1000, 0.40, 0.06);
  });

  test("desk penalty does NOT apply to non-workstation tasks", () => {
    const rng = mulberry32(13);
    let incomplete = 0;
    for (let i = 0; i < 1000; i++) {
      const r = rollOutcome({
        rng,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 0,
        taskType: "architecture",
        isUnassigned: true, // ignored — not a workstation task
      });
      if (r === "incomplete") incomplete++;
    }
    approx(incomplete / 1000, 0.15, 0.05);
  });
});

describe("rollOutcome — lesson bonus offsets incomplete", () => {
  test("20 lessons cancels 10% of incomplete rate (cap)", () => {
    const rngWith = mulberry32(17);
    const rngWithout = mulberry32(17);
    let incompleteWith = 0;
    let incompleteWithout = 0;
    for (let i = 0; i < 1000; i++) {
      const rWith = rollOutcome({
        rng: rngWith,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 0,
        lessonBonus: lessonSuccessBonus(20), // +0.10
      });
      const rWithout = rollOutcome({
        rng: rngWithout,
        precisionSkill: 15,
        creativitySkill: 0,
        reliabilitySkill: 0,
        lessonBonus: 0.0,
      });
      if (rWith === "incomplete") incompleteWith++;
      if (rWithout === "incomplete") incompleteWithout++;
    }
    // baseline 0.85 → 0.95 with bonus → 5% incomplete vs 15%
    approx(incompleteWith / 1000, 0.05, 0.04);
    approx(incompleteWithout / 1000, 0.15, 0.05);
  });

  test("lessonBonus is defensively clamped to 0.10", () => {
    // Force success-check to always pass by feeding a deterministic RNG that
    // returns < 1.0; the clamp protects against a caller passing a huge bonus.
    const rng: Rng = () => 0.94; // > 0.85 baseline, must rely on bonus to pass
    const r = rollOutcome({
      rng,
      precisionSkill: 15,
      creativitySkill: 0,
      reliabilitySkill: 0,
      lessonBonus: 999, // clamps to 0.10 → adjusted = 0.95
    });
    // 0.94 < 0.95 → success branch → bug check → clean
    assert.notEqual(r, "incomplete");
  });
});

describe("rollOutcome — known check-order", () => {
  test("first roll is reliability (incomplete path)", () => {
    // Force rng() > adjustedSuccess on first call.
    const seq = [0.99, 0.0, 0.0];
    let i = 0;
    const rng: Rng = () => seq[i++]!;
    const r = rollOutcome({
      rng,
      precisionSkill: 15,
      creativitySkill: 0,
      reliabilitySkill: 0, // baseline 0.85; 0.99 > 0.85 → incomplete
    });
    assert.equal(r, "incomplete");
  });

  test("second roll is bug (precision)", () => {
    // Skip reliability (yields success), then trigger bug.
    const seq = [0.0, 0.0, 0.0];
    let i = 0;
    const rng: Rng = () => seq[i++]!;
    const r = rollOutcome({
      rng,
      precisionSkill: 0, // bugChance 0.4 → 0.0 < 0.4 → bug
      creativitySkill: 0,
      reliabilitySkill: 15,
    });
    assert.equal(r, "bug");
  });

  test("third roll is crit (divergent only)", () => {
    const seq = [0.0, 0.99, 0.0];
    let i = 0;
    const rng: Rng = () => seq[i++]!;
    const r = rollOutcome({
      rng,
      precisionSkill: 15,
      creativitySkill: 50, // crit cap 1.0
      reliabilitySkill: 15,
      isDivergentTask: true,
    });
    assert.equal(r, "crit");
  });

  test("non-divergent skips crit roll → clean", () => {
    const seq = [0.0, 0.99, 0.0];
    let i = 0;
    const rng: Rng = () => seq[i++]!;
    const r = rollOutcome({
      rng,
      precisionSkill: 15,
      creativitySkill: 50,
      reliabilitySkill: 15,
      isDivergentTask: false,
    });
    assert.equal(r, "clean");
  });
});

describe("constants — Dart parity pins", () => {
  test("must equal Dart-side constants exactly", () => {
    // If any of these drift, update lib/services/task_outcome.dart in the
    // same commit. Keeping the values inline (not imported) is intentional.
    assert.equal(kUnassignedIncompletePenalty, 0.25);
    assert.equal(kMaxSpecializationCritBonusInRoll, 0.30);
    assert.equal(kMaxLessonSuccessBonusInRoll, 0.10);
    assert.equal(kMaxProjectMemoryBonusInRoll, 0.15);
  });
});

describe("mulberry32 — determinism", () => {
  test("same seed → same sequence", () => {
    const a = mulberry32(123);
    const b = mulberry32(123);
    for (let i = 0; i < 20; i++) {
      assert.equal(a(), b());
    }
  });
  test("different seeds → different sequences", () => {
    const a = mulberry32(123);
    const b = mulberry32(124);
    let differed = 0;
    for (let i = 0; i < 20; i++) {
      if (a() !== b()) differed++;
    }
    assert.ok(differed > 18, `differed=${differed}`);
  });
});
